src/ray/common/asio/instrumented_io_context.h

```diff
@@ -61,6 +61,33 @@ struct GuardedGlobalStats {
   mutable absl::Mutex mutex;
 };
 
+/// An opaque stats handle, used to manually instrument event loop handlers that don't
+/// hook into a .post() call.
+struct StatsHandle {
+  std::string handler_name;
+  int64_t start_time;
+  std::shared_ptr<GuardedHandlerStats> handler_stats;
+  std::shared_ptr<GuardedGlobalStats> global_stats;
+  bool execution_recorded = false;
+
+  StatsHandle(std::string handler_name_, int64_t start_time_,
+              std::shared_ptr<GuardedHandlerStats> handler_stats_,
+              std::shared_ptr<GuardedGlobalStats> global_stats_)
+      : handler_name(std::move(handler_name_)),
+        start_time(start_time_),
+        handler_stats(std::move(handler_stats_)),
+        global_stats(std::move(global_stats_)) {}
+
+  ~StatsHandle() {
+    if (!execution_recorded) {
+      // If handler execution was never recorded, we need to clean up some queueing
+      // stats in order to prevent those stats from leaking.
+      absl::MutexLock lock(&(handler_stats->mutex));
+      handler_stats->stats.curr_count--;
+    }
```

Comment:
I haven't been able to think of a possible race here yet, since the execution recording thread should hold a shared pointer to this struct, so destruction shouldn't take place when recording execution between modifying queueing stats and setting `handle->execution_recorded`.